src/eip712.rs

```diff
@@ -843,6 +981,237 @@ impl TypedData {
+    ComplexArrayValue(String),
+    #[error("Value must be boolean, number, string, array or struct. Property: {0}")]
+    ComplexValue(String),
+    #[error("Missing primaryType in recursive output. primaryType: {0}")]
+    MissingPrimaryTypeInRecursiveOutput(String),
+    #[error("JCS: {0}")]
+    JCS(serde_json::Error),
+    #[error("Proof type already exists")]
+    ProofAlreadyExists,
+}
+
+/// Generate EIP-712 types for EthereumEip712Signature2021
+///
+/// <https://w3c-ccg.github.io/ethereum-eip712-signature-2021-spec/#types-generation>
+/// from <https://github.com/w3c-ccg/ethereum-eip712-signature-2021-spec/pull/25>
+pub fn generate_types(
+    doc: &EIP712Value,
+    primary_type: Option<StructName>,
+) -> Result<HashMap<StructName, StructType>, TypesGenerationError> {
+    // 1
+    let mut output = HashMap::default();
+    // 2
+    // TypedDataField == MemberVariable
+    let mut types = StructType::default();
+    // 3
+    // Using JCS here probably has no effect:
+    // https://github.com/davidpdrsn/assert-json-diff
+    let doc_jcs = serde_jcs::to_string(doc).map_err(TypesGenerationError::JCS)?;
+    let doc: EIP712Value = serde_json::from_str(&doc_jcs).map_err(TypesGenerationError::JCS)?;
+    // 4
+    let primary_type = primary_type.unwrap_or_else(|| StructName::from("Document"));
+    // 5
+    let object = doc
+        .as_struct()
+        .ok_or(TypesGenerationError::ExpectedObject)?;
+    let mut props: Vec<(&String, &EIP712Value)> = object.iter().collect();
+    // Iterate through object properties in the order JCS would sort them.
+    // https://datatracker.ietf.org/doc/html/rfc8785#section-3.2.3
+    props.sort_by_cached_key(|(name, _value)| name.encode_utf16().collect::<Vec<u16>>());
+    for (property_name, value) in props {
+        match value {
+            // 6
+            EIP712Value::Bool(_) => {
+                // 6.1
+                types.0.push(MemberVariable {
+                    type_: EIP712Type::Bool,
+                    name: String::from(property_name),
+                });
+            }
+            EIP712Value::Integer(_) => {
+                // 6.2
+                types.0.push(MemberVariable {
+                    type_: EIP712Type::UintN(256),
+                    name: String::from(property_name),
+                });
+            }
+            EIP712Value::String(_) => {
+                // 6.3
+                types.0.push(MemberVariable {
+                    type_: EIP712Type::String,
+                    name: String::from(property_name),
+                });
+            }
+            // 7
+            EIP712Value::Array(array) => {
+                // Ensure values have same primitive type.
+                let mut values = array.into_iter();
+                let first_value = values
+                    .next()
+                    .ok_or_else(|| TypesGenerationError::EmptyArray(property_name.clone()))?;
+                match first_value {
```

Comment:
Considering there are only three acceptable variant, the validation could be done when deserialising.

Reply:
These constraints are not general to EIP-712 types, but only for this specific types generation algorithm. A new struct type could be made for the subset of EIP-712 that this types generation algorithm supports.